### DISCUSSION TITLE
RFC: proposal for REST proxy adapter

### DISCUSSION THREAD

**Discussion by twilson63** [hyper63/atlas#250] [Ideas]
This proposal is for a rest adapter for hyper, a kind of proxy adapter, the adapter would pass through to a REST integration, allowing for adapters to built using other technologies than Deno. For example, we could support a p2p database `hyperbee` which is tightly coupled to NodeJS, we could support hyperbee by creating an `hyper-adapter-rest` adapter that basically, takes the port methods and implements them with `fetch` rest calls. This will give us the ability to implement any service interface in any technology.

## Pros

* flexibility

## Cons

* performance - extra hop possibly
* should only be recommended for embeddable technologies managed tech adapter should be native to Deno.

**Comment by TillaTheHun0** [comment]
Cool idea. Is this an adapter, or is it merely an approach for building one? I could see a pretty generic rest "adapter" implementation like:

```js

const restAdapter = (methods = []) =>
  (url, secret) =>
    methods.reduce((acc, method) => ({
      ...acc,
      [method]: async (...args) => {
        const res = await fetch(url, {
          headers: {
            Authorization: `Bearer ${secret}`
          },
          body: JSON.stringify({
            method,
            payload: args
          })
        })

        return res.json()
      }
    }))

// then use
const adapter = restAdapter(
  ['createDatabase', 'removeDatabase', 'createDocument', '...']
)('https://my.rest.service', Deno.env.get('REST_ADAPTER_SECRET'))

```

Alternatively, we could have docs for how to do build this sort of thing.
